No Response on Initial Message problem fixed and sound feature added

amansoni7477030 opened the pull request:
1. Fxed a major issue with responses, as defined in https://github.com/FlowiseAI/Flowise/issues/2291 and in issue #158. Now, it will always show a response, even if the server is very slow with lower CPU and RAM. If the websocket takes time to connect and streaming is already over, it will directly show the result.
2. Added a new feature of sound: now users can add input and output sound effects for sending and receiving messages. Below are the parameters used:
```
sendMessageSound: true,
// sendSoundLocation: "send_message.mp3", // If not used, the default sound effect will play when sendMessageSound is true.
receiveMessageSound: true,
// receiveSoundLocation: "receive_message.mp3", // If not used, the default sound effect will play when receiveMessageSound is true.
```

HenryHengZJ commented:
Now we store the base64 string of the sound, it makes the `web.js` becomes extremely long: 
![image](https://github.com/FlowiseAI/FlowiseChatEmbed/assets/26460777/cfb704da-26cc-44f4-80cd-4a52b15746c7)

perhaps we could store it to like `raw.github` or some other storage, instead of local

HenryHengZJ commented on src/components/Bot.tsx:
```diff
@@ -429,9 +495,9 @@ export const Bot = (botProps: BotProps & { class?: string }) => {
+        updateLastMessage(text, data?.chatMessageId, data?.sourceDocuments, data?.fileAnnotations, data?.agentReasoning, data.text);
       } else {
-        updateLastMessage('', data?.chatMessageId, data?.sourceDocuments, data?.fileAnnotations, data?.agentReasoning);
+        updateLastMessage('', data?.chatMessageId, data?.sourceDocuments, data?.fileAnnotations, data?.agentReasoning, data.text);
```
when not streaming, do we have `data.text`?

amansoni7477030 replied:
yes

amansoni7477030 replied:
@HenryHengZJ In the code, there is a rare condition where even if the if (i === data.length - 1) part executes, sometimes there is no output or the UI remains blank. To address this issue, I added a condition to check if resultText is available but the UI hasn't updated (!uiUpdated). This ensures that the UI always displays the output. Please check the attached image where the first condition was executed but did not produce any output, hence the second condition was also executed.

amansoni7477030 commented:
> Now we store the base64 string of the sound, it makes the `web.js` becomes extremely long: ![image](https://private-user-images.githubusercontent.com/26460777/342507731-cfb704da-26cc-44f4-80cd-4a52b15746c7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTkyOTYzMTUsIm5iZiI6MTcxOTI5NjAxNSwicGF0aCI6Ii8yNjQ2MDc3Ny8zNDI1MDc3MzEtY2ZiNzA0ZGEtMjZjYy00NGY0LTgwY2QtNGE1MmIxNTc0NmM3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjI1VDA2MTMzNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE2NTY5ZDA1MDM5NTdhYTQ2MDI3NTJhNTJlYWU1YzdlMGE0MzI0ZTc2OTY3NTRkMmI2ZjI4YWM3YzhlNDlkZTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.x1daxY3vSisQ0vjO15NC3TwO7NBS_8RawEsUr0F6ocY)
> 
> perhaps we could store it to like `raw.github` or some other storage, instead of local

@HenryHengZJ Initially, I used GitHub Raw, which worked well. However, I started receiving a warning about "Third-party cookies being blocked in future Chrome versions as part of Privacy Sandbox." This means that after 2024, it may not work directly in Chrome, and we'll need to use some workarounds. Therefore, I reconsidered and decided that using CDN is the best option, although it can sometimes be slow. That's why I switched to using local storage. As you suggested, I've updated the code so that it now defaults to using CDN. If needed, you can modify the CDN link.
